samples/cpp/videocapture_with_audio.cpp

```diff
@@ -0,0 +1,146 @@
+/*
+ * This sample demonstrates how to read and display frames from a video file, a camera device or an image sequence.
+ * The sample also demonstrates extracting audio buffers from a video file.
+
+ * Capturing video and audio simultaneously from camera device is not supported by OpenCV.
+*/
+
+
+#include <opencv2/core.hpp>
+#include <opencv2/videoio.hpp>
+#include <opencv2/highgui.hpp>
+#include <iostream>
+
+using namespace cv;
+using namespace std;
+
+string keys =
+"{input i | | Path to input image or video file. Skip this argument to capture frames from a camera.}"
+"{with_audio wt | | set --with_audio if you want to extract audio buffers. If no input video file is provided only audio samples will be extracted}"
+"{help h | | print usage}";
+
+int main(int argc, char** argv)
+{
+    CommandLineParser parser(argc, argv, keys);
+
+    parser.about("This sample demonstrates how to read and display frames from a video file, a camera device or an image sequence \n"
+                 "The sample also demonstrates extracting audio buffers from a video file\n"
+                 "Capturing video and audio simultaneously from camera device is not supported by OpenCV\n"
+                 "Usage:\n --input/-i=<video file, image sequence> Skip this argument to capture frames from a camera\n"
+                 "--with_audio/-wt, set --with_audio if you want to extract audio buffers. If no input video file is provided only audio samples will be extracted\n"
+                 "q,Q,esc -- quit\n"
+                 "You can also pass the path to an image sequence and OpenCV will treat the sequence just like a video\n"
+                 "example: --input=right%02d.jpg\n");
+
+    // Check for 'help' argument to display usage information
+    if (parser.has("help"))
+    {
+        parser.printMessage();
+        return 0;
+    }
+
+    // Get input file path and check if audio extraction is required
+    string file = parser.get<string>("input");
+    bool isWithAudio = parser.has("with_audio");
+
+    Mat videoFrame;
+    Mat audioFrame;
+    vector<vector<Mat>> audioData;
+    VideoCapture cap;
+
+    // Setup VideoCapture based on user input
+    if (isWithAudio) {
+        if (file.empty()) {
+            // No input file, extract audio from default device
+            cout << "Extracting audio samples from default device..." << endl;
+            vector<int> micParams{ CAP_PROP_AUDIO_STREAM, 0, CAP_PROP_VIDEO_STREAM, -1, CAP_PROP_AUDIO_DATA_DEPTH, CV_16S };
+            cap.open(0, CAP_ANY, micParams);
+            if (!cap.isOpened()) {
+                cerr << "ERROR! Unable to open camera and microphone" << endl;
+                return -1;
+            }
```

Comment:
Something is wrong with microphone handling here. I get "core dumped" error on my Ubuntu 18.04 with web camera:
```
./bin/example_cpp_videocapture_with_audio --wt
Extracting audio samples from default device...
[ WARN:0@0.089] global cap_gstreamer.cpp:2840 handleMessage OpenCV | GStreamer warning: Embedded video playback halted; module v4l2src0 reported: Internal data stream error.
[ WARN:0@0.092] global cap_gstreamer.cpp:1698 open OpenCV | GStreamer warning: unable to start pipeline
[ WARN:0@0.092] global cap_gstreamer.cpp:1173 isPipelinePlaying OpenCV | GStreamer warning: GStreamer: pipeline have not been created
Audio Data Depth: <invalid depth>
Audio Samples Per Second: -1
Total Audio Channels: -1
Total Audio Streams: -1
terminate called after throwing an instance of 'std::length_error'
  what():  vector::_M_default_append
(core dumped)
```

Reply:
@asmorkalov It looks like audio device is only supported via MSMF backend. The seg fault can be handled and can work by adding a check on number of channels while resizing audioData, but audio is still not extracted

```suggestion
            cout << "Extracting audio samples from default device..." << endl;
            vector<int> micParams{ CAP_PROP_AUDIO_STREAM, 0, CAP_PROP_VIDEO_STREAM, -1, CAP_PROP_AUDIO_DATA_DEPTH, CV_16S };
            cap.open(0, MSMF, micParams); // Audio extraction only supported with MSMF backend via camera device
            if (!cap.isOpened()) {
                cerr << "ERROR! Unable to open camera and microphone" << endl;
                return -1;
            }
```

Reply:
1. GStreamer supports audio also. 
2. Let's add a check.